keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,553 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, a Kubernetes Service accepts a list of ports to be exposed by it.
+It is possible to specify any number of ports (as long as the service is within the max object size limit), by listing them in the service spec.
+This can be tedious if the service needs a large number of ports.
+This KEP proposes to add a new field to the ServicePort spec to allow exposing the entire port range(1 to 65535).
+
+## Motivation
+
+There are several applications like SIP apps or RTP which needs a lot of ports to run multiple calls or media streams.
+Currently the only option is to specify every port in the Service Spec. A request for port ranges in Services has been open in https://github.com/kubernetes/kubernetes/issues/23864. Implementing port ranges are challenging since iptables/ipvs do not support remapping port ranges. Hence, the proposal to set a single field in order to expose the entire port range and implement the service clients and endpoints accordingly.
+
+### Goals
+
+* Allow users to optionally expose the entire port range via a Service (of Type LoadBalancer or ClusterIP).
+
+### Non-Goals
+
+* Supporting Port Ranges in a Service.
+* Changing the default behavior of Service ports.
+
+## Proposal
+
+The proposal here is to introduce an “allPorts” boolean field to the [servicePort API.](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#serviceport-v1-core)
+This field is introduced within ServicePort rather than ServiceSpec, because a ServicePort entry is needed to infer the protocol.
+
+This field will be supported for Service Type=ClusterIP and ServiceType=LoadBalancer. This is not applicable to ExternalName services. NodePort services are not supported.
+A NodePort service accepts traffic on a given port on the Node, redirecting it to the specified targetPort of the service endpoints. Support AllPorts for a NodePort service means -
+traffic to any port on the node, will be forwarded to its endpoints on the same port. This could potentially break networking on the node, if traffic for, say, port 22 got forwarded to \<endpoint IP\>:22.
+
+In order to expose the entire port range on a supported service, a user needs to:
+1) Create a single ServicePort entry in the ServiceSpec and set "allPorts" to True.
```

Comment:
Curious if we've explored adding an `endPort` field, similar to what we added for port ranges for NetworkPolicy. Are there use-cases for _all ports_ or just a really large range of ports?

Reply:
I have assumed it is more "entire IP" than "all ports". I.e no protocol, just L3.

Reply:
The use-case is mostly got a large range of ports rather than allPorts. Exposing the entire port range + protocols (i.e making it an IP-level LoadBalancer service) simplifies some aspects around iptables rules, env variables for service ports, SRV records etc.

Reply:
Also, from a survey that was sent out a couple of months back, 70% of responders indicated that  IP-level loadbalancing solves their usecase - https://docs.google.com/presentation/d/1FO9H55-gnDh2RIqOZMDoP4OPVbaIaPhKl9C5to4uNNE/edit#slide=id.gdf6ff10943_0_56